apis/apiextensions/v1/composition_types.go

```diff
@@ -213,6 +222,16 @@ type Patch struct {
 	// +optional
 	FromFieldPath *string `json:"fromFieldPath,omitempty"`
 
+	// FromFieldPathStrategy determines how to patch from a field path. The
+	// default strategy is 'Optional', which means the patch will be a no-op if
+	// the specified fromFieldPath does not exist. Use the 'Required' strategy
+	// if the patch should fail if the specified fromFieldPath does not exist.
+	// The Required strategy can be useful to ensure that a composed resource
+	// is not rendered until the specified fromFieldPath is populated.
+	// +kubebuilder:validation:Enum=Optional;Required
+	// +optional
+	FromFieldPathStrategy *FromFieldPathStrategy `json:"fromFieldPathStrategy,omitempty"`
```

Comment:
This is kind of related to @benagricola question on https://github.com/crossplane/crossplane/pull/2134#discussion_r575077878, but have you thought about how this works patching back the other direction? Do we want to also error if the `FromFieldPath` on the `Composition` that we are trying to patch back to the XR / XRC is not present? I don't think it is necessarily a bad idea to error in that case, but it would have different implications because the managed resources could exist but we would still be erroring out.

Reply:
I'm not quite following you here.

A `ToCompositeFieldPath` type patch will patch from the composed resources (i.e. the managed resources) back to the XR - not from the `Composition`. This PR will therefore let you declare that the composed resource field path you're patching from is required to exist, just the same as if you were patching in the other direction. We're not changing a default or anything, so if you don't want the field to be required you can just stick with it being optional.

Reply:
Sorry when I said `Composition` I meant the managed resource that was rendered out. My point here is just that it is a different case when you are patching back to the XR and could be more difficult to debug than patching to managed resource. Maybe I am missing some context about when patching actually takes place when patching to the XR though. Is it before the managed resource actually gets created? Once again, not saying we shouldn't do this, just trying to understand the implications.

Reply:
@negz I spent some more time looking into this today and I believe I have confirmed by assumption about how this works. It appears that we will error out rendering composed managed resources in the case we have a missing required from field that is being patched back to the XR:
https://github.com/crossplane/crossplane/blob/5f239dbb7c399a8d544518be2be23ad16f98a71d/internal/controller/apiextensions/composite/reconciler.go#L471

I believe this means that we would not be able to do any required patching from the composed managed resources to the XR in the case where the from field was in the status or is late initialized. I guess we are just saying that these types of fields should not be marked as required patches, but it may be nice to instead say something like a resource cannot be considered ready until the field exists. Although I guess a readiness check should be used in that case. Anyway, you likely already considered this but may be worth documenting that it is a rare case when someone would want a required patch from the managed resource to the XR.

Reply:
Got it. The concern is that we might get stuck in a 'chicken and egg' situation in which we require a patch from a late initialized composed resource field, and thus we'll error out before we ever create the managed resource. I've pushed a commit that should address this - let me know if you think this needs more attention.

Reply:
@negz I think your commit has solved the primary issue, but my initial concern was actually around a more subtle one. Your commit solves the case where we would refuse to create _any_ resources, but I was thinking of the case where we just refuse to create _some_ resource because of an error with a missing required patch from a different managed resource to the XR. Let me outline the logic to see if I am understanding this correctly:
1. First we render the composed managed resources. Your commit addresses the case where we error out here from a missing from field that is required but is not yet populated.

https://github.com/crossplane/crossplane/blob/5f239dbb7c399a8d544518be2be23ad16f98a71d/internal/controller/apiextensions/composite/reconciler.go#L422

2. Next we go through a loop of those rendered managed resources.

https://github.com/crossplane/crossplane/blob/5f239dbb7c399a8d544518be2be23ad16f98a71d/internal/controller/apiextensions/composite/reconciler.go#L441

3. In that loop, we first `Apply` the rendered managed resource, then fetch connection details, then check if it is ready, then finally.. we render the composite based on any patches from the managed resource to the XR. This is where we can still error on a missing required from field.

https://github.com/crossplane/crossplane/blob/5f239dbb7c399a8d544518be2be23ad16f98a71d/internal/controller/apiextensions/composite/reconciler.go#L471

4. This makes sense, but importantly we don't just let the user know about that error, we _stop creating subsequent composed managed resources_. This means that order of the templates in a `Composition` now matters. If a managed resource with a required from field patch to the XR depends on (i.e. has a cross-resource reference to or just on the provider needs it to be created) a subsequent managed resource, that subsequent managed resource will never get created. For example, this is very contrived, but if we had a `Subnetwork` that references a `Network` and exists before that `Network` in a `Composition` and we had a required from field path patch from the `Subnetwork` to the `XR` on a field that gets populated after the `Subnetwork` is created successfully, we would never resolve because we would error out in the `Apply` loop on the missing from field, meaning the `Network` does not get created, so the `Subnetwork` will never have the field populated, and we will be stuck.

As mentioned before, this is a specific case, and it is not necessarily _wrong_, but it is a design decision. If we keep it as-is we should likely document a best practice of always having a resource that is dependent on another come after it in a `Composition` (users seem to already do this by default, but I don't think we don't specifically say to do so). Alternatively, it seems like we could do the patching from composed managed resources back to the XR in a separate loop (i.e. not coupled to the creation of the managed resources) and we would avoid this issue.

Reply:
Fair point.

My hunch is that required `ToCompositeFieldPath` patches will be rare or non-existent. The only use case that I'm aware of for required patches is to block the creation of the patched resource in order to ensure the `toFieldPath` isn't late initialised to an undesirable value.  e.g. We want to make sure a composed resource's `spec.forProvider.name` doesn't get defaulted to `default` if we're patching from a late-initialized XR field like `spec.claimRef.namespace`. Given that the XR will always exist already when we're patching from a composed resource I can't think of any reason someone would want to use `ToCompositeFieldPath`.

So I suspect that this case in which a required `ToCompositeFieldPath` patch prevents the XR from ever rendering successfully is fairly hypothetical. That said - your suggested fix of breaking up the apply-then-assess loop into multiple loops seems like a low-risk way to ensure we never need to worry about it.